### ISSUE TITLE
Removing containers doesn't remove the nodes from Consul. 

### ISSUE THREAD

**Issue by ashish235** [gliderlabs/docker-consul#108]
On removing or deleting the containers, the consul still keeps the records for those containers. Hence it is creating issues in my service discovery. 

Thanks,
Ashish

**Comment by djenriquez** [comment]
@ashish235, this is actually a feature for Consul. If you don't gracefully leave the cluster, by running `$ docker exec consul consul leave`, then the servers will keep the nodes in their discovery store. This is because the Consul servers do not know what happened to that node, only that it was available one minute, and then now it cannot communicate with it.

Consul will remove nodes that it has not received an acknowledgement from after three days. Otherwise, you can perform a curl command against a consul server to [force-leave](https://www.consul.io/docs/agent/http/agent.html) the node that is failing.

**Comment by ashish235** [comment]
@djenriquez , I think you misunderstood my issue. My problem is not about the consul node but the services which I 've registered in Consul. Say I 've a service A with port exposed 80 and 3000, two services A:80 and A:3000 gets added to Consul, becuase I 'm using registrator also. Now when I remove the containers, the services remain. Even if I stop the registrator, restart the consul, deregister all the services, the services come back.  Strange!

--
Ashish

**Comment by djenriquez** [comment]
Gotcha, so when you start service A and service B, they show up in consul. When you stop service A and service B, they remain in consul?

If this is the case, you're issue is most likely related to Registrator. Registrator is in charge or registering and deregistering containers to consul by monitoring the docker.sock.

What docker run command are you using to run Registrator? Also can you check the version of Registrator you are running? There was a release made to Registrator about two weeks ago that was pretty significant.

**Comment by ashish235** [comment]
@djenriquez, yeah got it. I 've updated the registrator now. The new services which I 'm adding now gets updated and removed as well. The stale services are still there, I can't remove them. :(

**Comment by djenriquez** [comment]
What you may want to do is call a force-removal of the node that those stale services reside in, then rejoin that node back to the cluster.

**Comment by hridyeshpant** [comment]
i am also facing same issue, i am using  using below command  in instance to register other docker container applications
docker run -d \
   --name=registrator \
    --net=host \
    --volume=/var/run/docker.sock:/tmp/docker.sock \
    gliderlabs/registrator:latest \
      consul://XX.XX.XX.XX:8500
Even i terminated the instance where all Registrator and other containers running,even through Consul server is showing service in UI with status "passing"

**Comment by philiphoy** [comment]
I expected that services registered when containers where added would be removed when the container was removed too. This does not appear to be the case unless a health check of some sort is used.

**Comment by hridyeshpant** [comment]
@philiphoy That is not happening . we are using ECS cluster having bunch of EC2 instance .
Each instance have consul agent and registrator container, but the problem is when we are updating our ECS cluster with new AMI published by AWS , old instance get terminated and replaced by new one. 
So all services which were running in old instances are still showing in consul server.

Any one have solution for this , do we really need to gracefully leave of docker registrator container before the instance get terminate?

@djenriquez 
is there any other solution without putting somethings for gracefully leave of docker registrator container  on termination?

**Comment by goverdhanr55** [comment]
I am facing a similar issue,  I have Version 1 of Service A running on port 9100.
I removed Version 1 and deployed Version 2 of Service A to the same port 9100, Consul shows both Version 1 and Version 2 of Service A running at port 9100.

This is messing up my service discovery.

Did anyone figure out a resolution?

**Comment by williamoneill** [comment]
I was having the same problem until I used Consul's _leave_on_terminate_ setting. Once the Consul Docker image shut down, the node left the Consul cluster and took its services with it. They never returned. Registrator was still running.

I'm using Consul with Registrator on an auto scaling ECS cluster.

BTW, if you are running Consul in Docker, like me, make sure it will respond to a SIGTERM signal. This involves putting brackets around [CMD in the Dockerfile](https://www.ctl.io/developers/blog/post/gracefully-stopping-docker-containers/).
